File: agents/python_sdk/clove_sdk/exceptions.py
"""Clove SDK exceptions.

Exception hierarchy for the Clove SDK providing typed error handling.
"""


class CloveError(Exception):
    """Base exception for all Clove SDK errors."""
    pass


class ConnectionError(CloveError):
    """Failed to connect to kernel or connection lost."""
    pass


class ProtocolError(CloveError):
    """Invalid protocol message, magic bytes, or malformed response."""
    pass


class TimeoutError(CloveError):
    """Operation timed out."""
    pass


class SyscallError(CloveError):
    """Syscall returned an error from the kernel.

    Attributes:
        opcode: The syscall opcode that failed
        details: Additional error details from kernel response
    """
    def __init__(self, message: str, opcode: int = None, details: dict = None):
        super().__init__(message)
        self.opcode = opcode
        self.details = details or {}


class PermissionDenied(SyscallError):
    """Agent lacks required permissions for the requested operation."""
    pass


class AgentNotFound(SyscallError):
    """Target agent does not exist or has terminated."""
    pass


class StateKeyNotFound(SyscallError):
    """Key not found in state store."""
    pass


class WorldNotFound(SyscallError):
    """World does not exist."""
    pass


class TunnelError(SyscallError):
    """Tunnel connection or relay error."""
    pass


class ValidationError(CloveError):
    """Invalid input parameters."""
    pass

File: sdk/typescript/src/exceptions.ts
/**
 * Exception classes for Clove SDK.
 */

import { SyscallOp } from "./protocol.js";

/**
 * Base class for all Clove SDK errors.
 */
export class CloveError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "CloveError";
  }
}

/**
 * Connection-related errors.
 */
export class ConnectionError extends CloveError {
  constructor(message: string) {
    super(message);
    this.name = "ConnectionError";
  }
}

/**
 * Protocol-related errors (invalid messages, etc).
 */
export class ProtocolError extends CloveError {
  constructor(message: string) {
    super(message);
    this.name = "ProtocolError";
  }
}

/**
 * Syscall execution errors.
 */
export class SyscallError extends CloveError {
  public readonly opcode?: SyscallOp;

  constructor(message: string, opcode?: SyscallOp) {
    super(message);
    this.name = "SyscallError";
    this.opcode = opcode;
  }
}

/**
 * Agent not found error.
 */
export class AgentNotFoundError extends SyscallError {
  constructor(message: string, opcode?: SyscallOp) {
    super(message, opcode);
    this.name = "AgentNotFoundError";
  }
}

/**
 * Validation error for invalid parameters.
 */
export class ValidationError extends CloveError {
  constructor(message: string) {
    super(message);
    this.name = "ValidationError";
  }
}

/**
 * Timeout error.
 */
export class TimeoutError extends CloveError {
  constructor(message: string = "Operation timed out") {
    super(message);
    this.name = "TimeoutError";
  }
}
